internal/wclayer/baselayerreader.go

```diff
@@ -0,0 +1,215 @@
+package wclayer
+
+import (
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/Microsoft/go-winio"
+	"github.com/Microsoft/hcsshim/internal/longpath"
+	"github.com/Microsoft/hcsshim/internal/oc"
+	"go.opencensus.io/trace"
+)
+
+type baseLayerReader struct {
+	s            *trace.Span
+	root         string
+	result       chan *fileEntry
+	proceed      chan bool
+	currentFile  *os.File
+	backupReader *winio.BackupFileReader
+}
+
+func newBaseLayerReader(root string, s *trace.Span) (r *baseLayerReader) {
+	r = &baseLayerReader{
+		s:       s,
+		root:    root,
+		result:  make(chan *fileEntry),
+		proceed: make(chan bool),
+	}
+	go r.walk()
+	return r
+}
+
+func (r *baseLayerReader) walkUntilCancelled() error {
+	root, err := longpath.LongAbs(r.root)
+	if err != nil {
+		return err
+	}
+
+	r.root = root
+
+	err = filepath.Walk(filepath.Join(r.root, filesPath), func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Indirect fix for https://github.com/moby/moby/issues/32838#issuecomment-343610048.
+		// Handle failure from what may be a golang bug in the conversion of
+		// UTF16 to UTF8 in files which are left in the recycle bin. Os.Lstat
+		// which is called by filepath.Walk will fail when a filename contains
+		// unicode characters. Skip the recycle bin regardless which is goodness.
+		if strings.EqualFold(path, filepath.Join(r.root, `Files\$Recycle.Bin`)) && info.IsDir() {
+			return filepath.SkipDir
+		}
+
+		r.result <- &fileEntry{path, info, nil}
+		if !<-r.proceed {
+			return errorIterationCanceled
+		}
+
+		return nil
+	})
+
+	if err == errorIterationCanceled {
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	utilityVMAbsPath := filepath.Join(r.root, utilityVMPath)
+	utilityVMFilesAbsPath := filepath.Join(r.root, utilityVMFilesPath)
+
+	// Ignore a UtilityVM without Files, that's not _really_ a UtiltyVM
+	if _, err = os.Lstat(utilityVMFilesAbsPath); err != nil {
+		if os.IsNotExist(err) {
+			return io.EOF
+		}
+		return err
+	}
+
+	err = filepath.Walk(utilityVMAbsPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if path != utilityVMAbsPath && path != utilityVMFilesAbsPath && !hasPathPrefix(path, utilityVMFilesAbsPath) {
+			if info.IsDir() {
+				return filepath.SkipDir
+			}
+			return nil
+		}
```

Comment:
_If_ I remember correctly, the `UtilityVM` folder can end up with other stuff in it after importing, and I didn't want to capture any of that, as it wasn't included in the MS-originated base layer tarballs. Basically, this lets us roundtrip a base layer tarball to disk and back with the only changes being due to some clear manually-output file entries, plus a difference in case-folding during sorting, and excludes files created by the `ImportBaseLayer` HCS call.

Specifically, I _think_ this was about not capturing backup registry hives or VHDX files.

Non-base layers don't have this issue, as the non-base layer exporter only exposes things we _want_ in the base layer out of the VHDX file, not the other stuff created on-disk by/for HCS, and there's no round-tripping tar->disk->tar possible anyway.

Looking at it, it might be clearer to manually emit an entry for `utilityVMAbsPath` and then walk from `utilityVMFilesAbsPath`? That should have the same effect, and would look more-similar to the way we export `Files/` earlier. (I suspect the MS-internal tool works this was, as I recall vaguely that `utilityVMAbsPath` in the upstream-provided tarballs has a `0000-00-00 00:00` timestamp and `--------` permission entry, suggesting it was a hand-crafted tar entry.

Reply:
Yeah that makes sense. I had forgotten about the VHD files. I don't have a preference here. We can keep this condition or create the `UtilityVM` directory manually, either is fine.